--- ../src-base/minecraft/net/minecraft/stats/StatisticsManager.java
+++ ../src-work/minecraft/net/minecraft/stats/StatisticsManager.java
@@ -4,6 +4,7 @@
 import java.util.Map;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.util.TupleIntJsonSerializable;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class StatisticsManager
 {
@@ -11,6 +12,10 @@
 
     public void increaseStat(EntityPlayer player, StatBase stat, int amount)
     {
+        org.bukkit.event.Cancellable cancellable = CraftEventFactory.handleStatisticsIncrease(player, stat, this.readStat(stat), amount);
+        if (cancellable != null && cancellable.isCancelled()) {
+            return;
+        }
         this.unlockAchievement(player, stat, this.readStat(stat) + amount);
     }
 
