--- ../src-base/minecraft/net/minecraftforge/server/terminalconsole/MinecraftFormattingConverter.java
+++ ../src-work/minecraft/net/minecraftforge/server/terminalconsole/MinecraftFormattingConverter.java
@@ -86,14 +86,14 @@
     private static final String LOOKUP = "0123456789abcdefklmnor";
 
     private static final String[] ansiCodes = new String[] {
-            "\u001B[0;30;22m", // Black §0
-            "\u001B[0;34;22m", // Dark Blue §1
-            "\u001B[0;32;22m", // Dark Green §2
-            "\u001B[0;36;22m", // Dark Aqua §3
-            "\u001B[0;31;22m", // Dark Red §4
-            "\u001B[0;35;22m", // Dark Purple §5
-            "\u001B[0;33;22m", // Gold §6
-            "\u001B[0;37;22m", // Gray §7
+            "\u001B[0;30m", // Black §0
+            "\u001B[0;34m", // Dark Blue §1
+            "\u001B[0;32m", // Dark Green §2
+            "\u001B[0;36m", // Dark Aqua §3
+            "\u001B[0;31m", // Dark Red §4
+            "\u001B[0;35m", // Dark Purple §5
+            "\u001B[0;33m", // Gold §6
+            "\u001B[0;37m", // Gray §7
             "\u001B[0;30;1m",  // Dark Gray §8
             "\u001B[0;34;1m",  // Blue §9
             "\u001B[0;32;1m",  // Green §a
@@ -158,16 +158,15 @@
         result.setLength(start + next);
 
         int pos = next;
-        int format;
-        do {
-            if (pos != next)
-            {
-                result.append(s, pos, next);
-            }
-
-            format = LOOKUP.indexOf(s.charAt(next + 1));
+        do
+        {
+            int format = LOOKUP.indexOf(Character.toLowerCase(s.charAt(next + 1)));
             if (format != -1)
             {
+                if (pos != next)
+                {
+                    result.append(s, pos, next);
+                }
                 if (ansi)
                 {
                     result.append(ansiCodes[format]);
