--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
@@ -7,6 +7,7 @@
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.IProjectile;
+import net.minecraft.entity.item.EntityEnderPearl;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Blocks;
 import net.minecraft.nbt.NBTTagCompound;
@@ -22,6 +23,7 @@
 import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public abstract class EntityThrowable extends Entity implements IProjectile
 {
@@ -31,8 +33,8 @@
     private Block inTile;
     protected boolean inGround;
     public int throwableShake;
-    protected EntityLivingBase thrower;
-    private String throwerName;
+    public EntityLivingBase thrower;
+    public String throwerName;
     private int ticksInGround;
     private int ticksInAir;
     public Entity ignoreEntity;
@@ -57,6 +59,7 @@
     {
         this(worldIn, throwerIn.posX, throwerIn.posY + (double)throwerIn.getEyeHeight() - 0.10000000149011612D, throwerIn.posZ);
         this.thrower = throwerIn;
+        this.projectileSource = (org.bukkit.entity.LivingEntity) throwerIn.getBukkitEntity(); // CraftBukkit
     }
 
     protected void entityInit()
@@ -196,7 +199,7 @@
                 {
                     flag = true;
                 }
-                else if (this.thrower != null && this.ticksExisted < 2 && this.ignoreEntity == null)
+                else if (this.thrower != null && this.ticksExisted < 2 && this.ignoreEntity == null && this.thrower == entity1) // CraftBukkit - MC-88491
                 {
                     this.ignoreEntity = entity1;
                     flag = true;
@@ -238,6 +241,15 @@
             raytraceresult = new RayTraceResult(entity);
         }
 
+        // Paper start - Call ProjectileCollideEvent
+        if (raytraceresult != null && raytraceresult.entityHit != null) {
+            com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileCollideEvent(this, raytraceresult);
+            if (event.isCancelled()) {
+                raytraceresult = null;
+            }
+        }
+        // Paper end
+
         if (raytraceresult != null)
         {
             if (raytraceresult.typeOfHit == RayTraceResult.Type.BLOCK && this.world.getBlockState(raytraceresult.getBlockPos()).getBlock() == Blocks.PORTAL)
@@ -247,6 +259,9 @@
             else if (!net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onImpact(raytraceresult);
+                if (this.isDead) {
+                    CraftEventFactory.callProjectileHitEvent(this, raytraceresult);
+                }
             }
         }
 
@@ -366,7 +381,7 @@
     {
         if (this.thrower == null && this.throwerName != null && !this.throwerName.isEmpty())
         {
-            this.thrower = this.world.getPlayerEntityByName(this.throwerName);
+            if (this instanceof EntityEnderPearl && this.world != null && this.world.paperConfig.disableEnderpearlExploit) { this.thrower = null; } // Paper - Don't store shooter name for pearls to block enderpearl travel exploit            this.thrower = this.world.getPlayerEntityByName(this.throwerName);
 
             if (this.thrower == null && this.world instanceof WorldServer)
             {
