--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityEvokerFangs.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityEvokerFangs.java
@@ -12,6 +12,7 @@
 import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class EntityEvokerFangs extends Entity
 {
@@ -136,7 +137,9 @@
         {
             if (entitylivingbase == null)
             {
+                CraftEventFactory.entityDamage = this;
                 p_190551_1_.attackEntityFrom(DamageSource.MAGIC, 6.0F);
+                CraftEventFactory.entityDamage = null;
             }
             else
             {
