--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/FMLRelaunchLog.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/FMLRelaunchLog.java
@@ -20,6 +20,7 @@
 package net.minecraftforge.fml.relauncher;
 
 import net.minecraftforge.fml.common.FMLLog;
+import net.minecraftforge.fml.common.MagmaFMLLogInjector;
 import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
 
 import org.apache.logging.log4j.Level;
@@ -45,7 +46,7 @@
 
     public static void log(String targetLog, Level level, String format, Object... data)
     {
-        LogManager.getLogger(targetLog).log(level, String.format(format, data));
+        new MagmaFMLLogInjector(LogManager.getLogger(targetLog)).log(level, String.format(format, data));
     }
 
     public static void log(Level level, String format, Object... data)
@@ -55,7 +56,7 @@
 
     public static void log(String targetLog, Level level, Throwable ex, String format, Object... data)
     {
-        LogManager.getLogger(targetLog).log(level, String.format(format, data), ex);
+        new MagmaFMLLogInjector(LogManager.getLogger(targetLog)).log(level, String.format(format, data), ex);
     }
 
     public static void log(Level level, Throwable ex, String format, Object... data)
