--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/ServerLaunchWrapper.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/ServerLaunchWrapper.java
@@ -19,21 +19,69 @@
 
 package net.minecraftforge.fml.relauncher;
 
+import org.magmafoundation.magma.configuration.MagmaConfig;
+import org.magmafoundation.magma.downloads.DownloadServerFiles;
+import org.magmafoundation.magma.downloads.MagmaUpdater;
+
 import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.core.LoggerContext;
-
 public class ServerLaunchWrapper {
 
+    public static long beginTime;
+    public static List<String> LAUNCH_ARGS;
+
     /**
      * @param args
      */
     public static void main(String[] args)
     {
-        new ServerLaunchWrapper().run(args);
+      System.out.println("Starting Magma");
+      LAUNCH_ARGS = Arrays.asList(args);
+
+        if(getJavaVersion() > 8){
+            try {
+                throw new JavaToNewException();
+            } catch (JavaToNewException e) {
+                e.printStackTrace();
+                System.out.println("Magma Only Support's Java 8");
+                return;
+            }
+        }
+
+      DownloadServerFiles.downloadMinecraftServer();
+
+      if(DownloadServerFiles.getLibrariesVersion()){
+          DownloadServerFiles.downloadServerLibraries();
+      }
+
+      if(!LAUNCH_ARGS.contains("-dau")) {
+          System.out.println("Checking for new version...");
+          MagmaUpdater magmaUpdater = new MagmaUpdater();
+          if (magmaUpdater.versionChecker() && MagmaConfig.instance.magmaAutoUpdater.getValues()) {
+              magmaUpdater.downloadJar();
+          }
+      }
+
+      new ServerLaunchWrapper().run(args);
     }
 
+    private static class JavaToNewException extends RuntimeException {
+
+    }
+
+    public static int getJavaVersion() {
+        String version = System.getProperty("java.version");
+        if (version.startsWith("1.")) {
+            version = version.substring(2);
+        }
+
+        int dotPos = version.indexOf('.');
+        int dashPos = version.indexOf('-');
+        return Integer.parseInt(version.substring(0, dotPos > -1 ? dotPos : dashPos > -1 ? dashPos : 1));
+    }
+
     private ServerLaunchWrapper()
     {
 
@@ -56,9 +104,9 @@
         {
             System.err.printf("We appear to be missing one or more essential library files.\n" +
             		"You will need to add them to your server before FML and Forge will run successfully.");
-            e.printStackTrace(System.err);
-            System.exit(1);
+
         }
+        beginTime = System.nanoTime();
 
         try
         {
