--- ../src-base/minecraft/net/minecraft/scoreboard/ScoreboardSaveData.java
+++ ../src-work/minecraft/net/minecraft/scoreboard/ScoreboardSaveData.java
@@ -222,6 +222,7 @@
         for (ScorePlayerTeam scoreplayerteam : this.scoreboard.getTeams())
         {
             NBTTagCompound nbttagcompound = new NBTTagCompound();
+            if (!com.destroystokyo.paper.PaperConfig.saveEmptyScoreboardTeams && scoreplayerteam.getDisplayName().isEmpty()) continue; // Paper
             nbttagcompound.setString("Name", scoreplayerteam.getName());
             nbttagcompound.setString("DisplayName", scoreplayerteam.getDisplayName());
 
