--- ../src-base/minecraft/net/minecraft/tileentity/TileEntitySign.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntitySign.java
@@ -3,6 +3,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.command.CommandException;
 import net.minecraft.command.CommandResultStats;
+import net.minecraft.command.ICommandListener;
 import net.minecraft.command.ICommandSender;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayer;
@@ -19,14 +20,19 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.command.CraftBlockCommandSender;
+import org.bukkit.craftbukkit.v1_12_R1.command.ProxiedNativeCommandSender;
+import org.bukkit.craftbukkit.v1_12_R1.util.CraftChatMessage;
 
 public class TileEntitySign extends TileEntity
 {
     public final ITextComponent[] signText = new ITextComponent[] {new TextComponentString(""), new TextComponentString(""), new TextComponentString(""), new TextComponentString("")};
     public int lineBeingEdited = -1;
-    private boolean isEditable = true;
+    public boolean isEditable = true;
     private EntityPlayer player;
     private final CommandResultStats stats = new CommandResultStats();
+    private static final boolean CONVERT_LEGACY_SIGNS = Boolean.getBoolean("convertLegacySigns");
+    public java.util.UUID signEditor; // Paper
 
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
@@ -38,6 +44,10 @@
             compound.setString("Text" + (i + 1), s);
         }
 
+        if (CONVERT_LEGACY_SIGNS) { // Paper
+            compound.setBoolean("Bukkit.isConverted", true);
+        }
+
         this.stats.writeStatsToNBT(compound);
         return compound;
     }
@@ -51,7 +61,7 @@
     {
         this.isEditable = false;
         super.readFromNBT(compound);
-        ICommandSender icommandsender = new ICommandSender()
+        ICommandSender icommandsender = new ISignCommandListener() // Paper
         {
             public String getName()
             {
@@ -61,6 +71,10 @@
             {
                 return permLevel <= 2; //Forge: Fixes  MC-75630 - Exploit with signs and command blocks
             }
+            public boolean canUseCommand(int permLevel, String commandName, String perm)
+            {
+                return permLevel <= 2; //Forge: Fixes  MC-75630 - Exploit with signs and command blocks
+            }
             public BlockPos getPosition()
             {
                 return TileEntitySign.this.pos;
@@ -79,19 +93,38 @@
             }
         };
 
+        // CraftBukkit start - Add an option to convert signs correctly
+        // This is done with a flag instead of all the time because
+        // we have no way to tell whether a sign is from 1.7.10 or 1.8
+
+        boolean oldSign = Boolean.getBoolean("convertLegacySigns") && !compound.getBoolean("Bukkit.isConverted");
+
         for (int i = 0; i < 4; ++i)
         {
             String s = compound.getString("Text" + (i + 1));
-            ITextComponent itextcomponent = ITextComponent.Serializer.jsonToComponent(s);
+            // ITextComponent itextcomponent = ITextComponent.Serializer.jsonToComponent(s);
+            if (s != null && s.length() > 2048) {
+                s = "\"\"";
+            }
 
             try
             {
-                this.signText[i] = TextComponentUtils.processComponent(icommandsender, itextcomponent, (Entity)null);
+                // ITextComponent ichatbasecomponent = ITextComponent.Serializer.jsonToComponent(s);  // Paper - move down - the old format might throw a json error
+
+                if (oldSign && !isLoadingStructure ){ // Paper - saved structures will be in the new format, but will not have isConverted
+                    signText[i] = CraftChatMessage.fromString(s)[0];
+                    continue;
+                }
+                // CraftBukkit end
+                ITextComponent ichatbasecomponent = ITextComponent.Serializer.jsonToComponent(s);  // Paper - after old sign
+                try {
+                    this.signText[i] = TextComponentUtils.processComponent(icommandsender, ichatbasecomponent, (Entity) null);
+                } catch (CommandException commandexception) {
+                    this.signText[i] = ichatbasecomponent;
+                }
+            } catch (com.google.gson.JsonParseException jsonparseexception) {
+                this.signText[i] = new TextComponentString(s);
             }
-            catch (CommandException var7)
-            {
-                this.signText[i] = itextcomponent;
-            }
         }
 
         this.stats.readStatsFromNBT(compound);
@@ -131,7 +164,10 @@
 
     public void setPlayer(EntityPlayer playerIn)
     {
-        this.player = playerIn;
+        // Paper start
+        //this.player = playerIn;
+        signEditor = playerIn != null ? playerIn.getUniqueID() : null;
+        // Paper end
     }
 
     public EntityPlayer getPlayer()
@@ -141,7 +177,7 @@
 
     public boolean executeCommand(final EntityPlayer playerIn)
     {
-        ICommandSender icommandsender = new ICommandSender()
+        ICommandSender icommandsender = new ISignCommandListener() // Paper
         {
             public String getName()
             {
@@ -158,6 +194,11 @@
             {
                 return permLevel <= 2;
             }
+
+            public boolean canUseCommand(int permLevel, String commandName, String perm)
+            {
+                return permLevel <= 2;
+            }
             public BlockPos getPosition()
             {
                 return TileEntitySign.this.pos;
@@ -187,7 +228,7 @@
             }
             public MinecraftServer getServer()
             {
-                return playerIn.getServer();
+                return MinecraftServer.getServerInstance(); // Paper - world maybe null
             }
         };
 
@@ -201,7 +242,12 @@
 
                 if (clickevent.getAction() == ClickEvent.Action.RUN_COMMAND)
                 {
-                    playerIn.getServer().getCommandManager().executeCommand(icommandsender, clickevent.getValue());
+                    // playerIn.getServer().getCommandManager().executeCommand(icommandsender, clickevent.getValue());
+                    CommandBlockBaseLogic.executeSafely(icommandsender, new ProxiedNativeCommandSender(
+                            icommandsender,
+                            new CraftBlockCommandSender(icommandsender),
+                            playerIn.getBukkitEntity()
+                    ), clickevent.getValue());
                 }
             }
         }
@@ -213,4 +259,5 @@
     {
         return this.stats;
     }
+    public interface ISignCommandListener extends ICommandSender {} // Paper
 }
