--- ../src-base/minecraft/net/minecraft/inventory/SlotFurnaceOutput.java
+++ ../src-work/minecraft/net/minecraft/inventory/SlotFurnaceOutput.java
@@ -4,7 +4,11 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.FurnaceRecipes;
+import net.minecraft.tileentity.TileEntityFurnace;
 import net.minecraft.util.math.MathHelper;
+import org.bukkit.entity.Player;
+import org.bukkit.event.inventory.FurnaceExtractEvent;
+import org.bukkit.craftbukkit.v1_12_R1.util.CraftMagicNumbers;
 
 public class SlotFurnaceOutput extends Slot
 {
@@ -70,11 +74,26 @@
                 i = j;
             }
 
+            // Magma start - validate inventory before attempting to cast it
+            if (this.inventory instanceof TileEntityFurnace) {
+                // CraftBukkit - fire FurnaceExtractEvent
+                Player player = (Player) this.player.getBukkitEntity();
+                TileEntityFurnace furnace = ((TileEntityFurnace) this.inventory);
+                org.bukkit.block.Block block = this.player.world.getWorld().getBlockAt(furnace.getPos().getX(), furnace.getPos().getY(), furnace.getPos().getZ());
+
+                if (removeCount != 0) {
+                    FurnaceExtractEvent event = new FurnaceExtractEvent(player, block, CraftMagicNumbers.getMaterial(stack.getItem()), removeCount, i);
+                    this.player.world.getServer().getPluginManager().callEvent(event);
+                    i = event.getExpToDrop();
+                }
+            }
+            // Magma end
+
             while (i > 0)
             {
                 int k = EntityXPOrb.getXPSplit(i);
                 i -= k;
-                this.player.world.spawnEntity(new EntityXPOrb(this.player.world, this.player.posX, this.player.posY + 0.5D, this.player.posZ + 0.5D, k));
+                this.player.world.spawnEntity(new EntityXPOrb(this.player.world, this.player.posX, this.player.posY + 0.5D, this.player.posZ + 0.5D, k, org.bukkit.entity.ExperienceOrb.SpawnReason.FURNACE, player)); // Paper
             }
         }
 
