--- ../src-base/minecraft/net/minecraft/pathfinding/PathNavigate.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/PathNavigate.java
@@ -1,6 +1,9 @@
 package net.minecraft.pathfinding;
 
 import javax.annotation.Nullable;
+
+import com.destroystokyo.paper.MCUtil;
+import com.destroystokyo.paper.event.entity.EntityPathfindEvent;
 import net.minecraft.block.Block;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
@@ -18,6 +21,13 @@
 public abstract class PathNavigate
 {
     protected EntityLiving entity;
+
+    // Paper start
+    public EntityLiving getEntity() {
+        return entity;
+    }
+    // Paper end
+
     protected World world;
     @Nullable
     protected Path currentPath;
@@ -99,6 +109,7 @@
         }
         else
         {
+            if(!new EntityPathfindEvent(getEntity().getBukkitEntity(), MCUtil.toLocation(getEntity().world, pos), null).callEvent()) { return null; } // Paper
             this.targetPos = pos;
             float f = this.getPathSearchRange();
             this.world.profiler.startSection("pathfind");
@@ -128,6 +139,7 @@
             }
             else
             {
+                if(!new EntityPathfindEvent(getEntity().getBukkitEntity(), MCUtil.toLocation(getEntity().world, blockpos), null).callEvent()) { return null; } // Paper
                 this.targetPos = blockpos;
                 float f = this.getPathSearchRange();
                 this.world.profiler.startSection("pathfind");
