--- ../src-base/minecraft/net/minecraft/world/border/WorldBorder.java
+++ ../src-work/minecraft/net/minecraft/world/border/WorldBorder.java
@@ -6,6 +6,7 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
@@ -24,6 +25,9 @@
     private int warningTime;
     private int warningDistance;
 
+    // TODO: Check if it's initialized properly
+    public WorldServer world;
+
     public WorldBorder()
     {
         this.endDiameter = this.startDiameter;
@@ -34,6 +38,21 @@
         this.warningDistance = 5;
     }
 
+    // Paper start
+    private final BlockPos.MutableBlockPos mutPos = new BlockPos.MutableBlockPos();
+
+    public boolean isBlockInBounds(int chunkX, int chunkZ) {
+        mutPos.setPos(chunkX, 64, chunkZ);
+        return isInBounds(mutPos);
+    }
+
+    public boolean isChunkInBounds(int chunkX, int chunkZ) {
+        mutPos.setPos(((chunkX << 4) + 15), 64, (chunkZ << 4) + 15);
+        return isInBounds(mutPos);
+    }
+    // Paper end
+
+    public boolean isInBounds(BlockPos blockposition) { return contains(blockposition); } // Paper - OBFHELPER
     public boolean contains(BlockPos pos)
     {
         return (double)(pos.getX() + 1) > this.minX() && (double)pos.getX() < this.maxX() && (double)(pos.getZ() + 1) > this.minZ() && (double)pos.getZ() < this.maxZ();
@@ -206,6 +225,7 @@
 
     public void addListener(IBorderListener listener)
     {
+        if (listeners.contains(listener)) return;
         this.listeners.add(listener);
     }
 
