--- ../src-base/minecraft/net/minecraft/world/gen/ChunkGeneratorOverworld.java
+++ ../src-work/minecraft/net/minecraft/world/gen/ChunkGeneratorOverworld.java
@@ -206,44 +206,37 @@
         this.biomesForGeneration = this.world.getBiomeProvider().getBiomes(this.biomesForGeneration, x * 16, z * 16, 16, 16);
         this.replaceBiomeBlocks(x, z, chunkprimer, this.biomesForGeneration);
 
-        if (this.settings.useCaves)
-        {
+        if (this.settings.useCaves && this.world.paperConfig.generateCaves) { // Paper
             this.caveGenerator.generate(this.world, x, z, chunkprimer);
         }
 
-        if (this.settings.useRavines)
-        {
+        if (this.settings.useRavines && this.world.paperConfig.generateCanyon) { // Paper
             this.ravineGenerator.generate(this.world, x, z, chunkprimer);
         }
 
         if (this.mapFeaturesEnabled)
         {
-            if (this.settings.useMineShafts)
-            {
+            if (this.settings.useMineShafts && this.world.paperConfig.generateMineshaft) { // Paper
                 this.mineshaftGenerator.generate(this.world, x, z, chunkprimer);
             }
 
-            if (this.settings.useVillages)
-            {
+            if (this.settings.useVillages && this.world.paperConfig.generateVillage) { // Paper
                 this.villageGenerator.generate(this.world, x, z, chunkprimer);
             }
 
-            if (this.settings.useStrongholds)
-            {
+            if (this.settings.useStrongholds && this.world.paperConfig.generateStronghold) { // Paper
                 this.strongholdGenerator.generate(this.world, x, z, chunkprimer);
             }
 
-            if (this.settings.useTemples)
-            {
+            if (this.settings.useTemples && this.world.paperConfig.generateTemple) { // Paper
                 this.scatteredFeatureGenerator.generate(this.world, x, z, chunkprimer);
             }
 
-            if (this.settings.useMonuments)
-            {
+            if (this.settings.useMonuments && this.world.paperConfig.generateMonument) { // Paper
                 this.oceanMonumentGenerator.generate(this.world, x, z, chunkprimer);
             }
 
-            if (this.settings.useMansions)
+            if (this.settings.useMansions && this.world.paperConfig.generateWoodlandMansion) // Paper
             {
                 this.woodlandMansionGenerator.generate(this.world, x, z, chunkprimer);
             }
@@ -296,6 +289,12 @@
                             f6 = 1.0F + f6 * 4.0F;
                         }
 
+                        // CraftBukkit start - fix MC-54738
+                        if (f5 < -1.8F) {
+                            f5 = -1.8F;
+                        }
+                        // CraftBukkit end
+
                         float f7 = this.biomeWeights[j1 + 2 + (k1 + 2) * 5] / (f5 + 2.0F);
 
                         if (biome1.getBaseHeight() > biome.getBaseHeight())
@@ -396,33 +395,27 @@
 
         if (this.mapFeaturesEnabled)
         {
-            if (this.settings.useMineShafts)
-            {
+            if (this.settings.useMineShafts && this.world.paperConfig.generateMineshaft) { // Paper
                 this.mineshaftGenerator.generateStructure(this.world, this.rand, chunkpos);
             }
 
-            if (this.settings.useVillages)
-            {
+            if (this.settings.useVillages && this.world.paperConfig.generateVillage) { // Paper
                 flag = this.villageGenerator.generateStructure(this.world, this.rand, chunkpos);
             }
 
-            if (this.settings.useStrongholds)
-            {
+            if (this.settings.useStrongholds && this.world.paperConfig.generateStronghold) { // Paper
                 this.strongholdGenerator.generateStructure(this.world, this.rand, chunkpos);
             }
 
-            if (this.settings.useTemples)
-            {
+            if (this.settings.useTemples && this.world.paperConfig.generateTemple) { // Paper
                 this.scatteredFeatureGenerator.generateStructure(this.world, this.rand, chunkpos);
             }
 
-            if (this.settings.useMonuments)
-            {
+            if (this.settings.useMonuments && this.world.paperConfig.generateMonument) { // Paper
                 this.oceanMonumentGenerator.generateStructure(this.world, this.rand, chunkpos);
             }
 
-            if (this.settings.useMansions)
-            {
+            if (this.settings.useMansions && this.world.paperConfig.generateWoodlandMansion) { // Paper
                 this.woodlandMansionGenerator.generateStructure(this.world, this.rand, chunkpos);
             }
         }
@@ -449,7 +442,7 @@
             }
         }
 
-        if (this.settings.useDungeons)
+        if (this.settings.useDungeons && this.world.paperConfig.generateDungeon) // Paper
         if (net.minecraftforge.event.terraingen.TerrainGen.populate(this, this.world, this.rand, x, z, flag, net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.DUNGEON))
         {
             for (int j2 = 0; j2 < this.settings.dungeonChance; ++j2)
@@ -594,33 +587,27 @@
     {
         if (this.mapFeaturesEnabled)
         {
-            if (this.settings.useMineShafts)
-            {
+            if (this.settings.useMineShafts && this.world.paperConfig.generateMineshaft) { // Paper
                 this.mineshaftGenerator.generate(this.world, x, z, (ChunkPrimer)null);
             }
 
-            if (this.settings.useVillages)
-            {
+            if (this.settings.useVillages && this.world.paperConfig.generateVillage) { // Paper
                 this.villageGenerator.generate(this.world, x, z, (ChunkPrimer)null);
             }
 
-            if (this.settings.useStrongholds)
-            {
+            if (this.settings.useStrongholds && this.world.paperConfig.generateStronghold) { // Paper
                 this.strongholdGenerator.generate(this.world, x, z, (ChunkPrimer)null);
             }
 
-            if (this.settings.useTemples)
-            {
+            if (this.settings.useTemples && this.world.paperConfig.generateTemple) { // Paper
                 this.scatteredFeatureGenerator.generate(this.world, x, z, (ChunkPrimer)null);
             }
 
-            if (this.settings.useMonuments)
-            {
+            if (this.settings.useMonuments  && this.world.paperConfig.generateMonument) { // Paper
                 this.oceanMonumentGenerator.generate(this.world, x, z, (ChunkPrimer)null);
             }
 
-            if (this.settings.useMansions)
-            {
+            if (this.settings.useMansions && this.world.paperConfig.generateWoodlandMansion) { // Paper
                 this.woodlandMansionGenerator.generate(this.world, x, z, (ChunkPrimer)null);
             }
         }
