--- ../src-base/minecraft/net/minecraftforge/fml/common/asm/FMLSanityChecker.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/asm/FMLSanityChecker.java
@@ -45,9 +45,9 @@
 
 public class FMLSanityChecker implements IFMLCallHook
 {
-    private static final String FMLFINGERPRINT =   "51:0A:FB:4C:AF:A4:A0:F2:F5:CF:C5:0E:B4:CC:3C:30:24:4A:E3:8E".toLowerCase().replace(":", "");
-    private static final String FORGEFINGERPRINT = "E3:C3:D5:0C:7C:98:6D:F7:4C:64:5C:0A:C5:46:39:74:1C:90:A5:57".toLowerCase().replace(":", "");
-    private static final String MCFINGERPRINT =    "CD:99:95:96:56:F7:53:DC:28:D8:63:B4:67:69:F7:F8:FB:AE:FC:FC".toLowerCase().replace(":", "");
+    private static final String FMLFINGERPRINT =   "1D:1A:4F:D5:C6:D3:95:91:E9:4E:78:9C:8C:4C:8D:72:54:98:D8:BE".toLowerCase().replace(":", "");
+    private static final String FORGEFINGERPRINT = "1D:1A:4F:D5:C6:D3:95:91:E9:4E:78:9C:8C:4C:8D:72:54:98:D8:BE".toLowerCase().replace(":", "");
+    private static final String MCFINGERPRINT =    "1D:1A:4F:D5:C6:D3:95:91:E9:4E:78:9C:8C:4C:8D:72:54:98:D8:BE".toLowerCase().replace(":", "");
     private LaunchClassLoader cl;
     private boolean liveEnv;
     public static File fmlLocation;
@@ -171,7 +171,7 @@
         }
         if (!goodFML)
         {
-            FMLLog.log.error("FML appears to be missing any signature data. This is not a good thing");
+            //FMLLog.log.error("FML appears to be missing any signature data. This is not a good thing");
         }
         return null;
     }
