--- ../src-base/minecraft/net/minecraft/entity/monster/EntityEnderman.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityEnderman.java
@@ -1,13 +1,11 @@
 package net.minecraft.entity.monster;
 
+import com.destroystokyo.paper.event.entity.EndermanAttackPlayerEvent;
+import com.destroystokyo.paper.event.entity.EndermanEscapeEvent;
 import com.google.common.base.Function;
 import com.google.common.base.Optional;
 import com.google.common.base.Predicate;
 import com.google.common.collect.Sets;
-import java.util.Random;
-import java.util.Set;
-import java.util.UUID;
-import javax.annotation.Nullable;
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
@@ -15,14 +13,7 @@
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.SharedMonsterAttributes;
-import net.minecraft.entity.ai.EntityAIAttackMelee;
-import net.minecraft.entity.ai.EntityAIBase;
-import net.minecraft.entity.ai.EntityAIHurtByTarget;
-import net.minecraft.entity.ai.EntityAILookIdle;
-import net.minecraft.entity.ai.EntityAINearestAttackableTarget;
-import net.minecraft.entity.ai.EntityAISwimming;
-import net.minecraft.entity.ai.EntityAIWanderAvoidWater;
-import net.minecraft.entity.ai.EntityAIWatchClosest;
+import net.minecraft.entity.ai.*;
 import net.minecraft.entity.ai.attributes.AttributeModifier;
 import net.minecraft.entity.ai.attributes.IAttributeInstance;
 import net.minecraft.entity.player.EntityPlayer;
@@ -35,11 +26,7 @@
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.pathfinding.PathNodeType;
-import net.minecraft.util.DamageSource;
-import net.minecraft.util.EntityDamageSourceIndirect;
-import net.minecraft.util.EnumParticleTypes;
-import net.minecraft.util.ResourceLocation;
-import net.minecraft.util.SoundEvent;
+import net.minecraft.util.*;
 import net.minecraft.util.datafix.DataFixer;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
@@ -47,7 +34,17 @@
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.craftbukkit.v1_12_R1.entity.CraftEnderman;
+import org.bukkit.entity.Enderman;
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
+import javax.annotation.Nullable;
+import java.util.Random;
+import java.util.Set;
+import java.util.UUID;
+
 public class EntityEnderman extends EntityMob
 {
     private static final UUID ATTACKING_SPEED_BOOST_ID = UUID.fromString("020E0DFB-87AE-4653-9556-831010E291A0");
@@ -97,25 +94,38 @@
 
     public void setAttackTarget(@Nullable EntityLivingBase entitylivingbaseIn)
     {
-        super.setAttackTarget(entitylivingbaseIn);
+        // CraftBukkit start - fire event
+        setAttackTarget(entitylivingbaseIn, EntityTargetEvent.TargetReason.UNKNOWN, true);
+    }
+
+    // Paper start
+    private boolean tryEscape(EndermanEscapeEvent.Reason reason){
+        return new EndermanEscapeEvent((CraftEnderman) this.getBukkitEntity(), reason).callEvent();
+    }
+    // Paper end
+
+    @Override
+    public boolean setAttackTarget(@Nullable EntityLivingBase entityliving, EntityTargetEvent.TargetReason reason, boolean fireEvent) {
+        if (!super.setAttackTarget(entityliving, reason, fireEvent)) {
+            return false;
+        }
+        entityliving = getAttackTarget();
+        // CraftBukkit end
         IAttributeInstance iattributeinstance = this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED);
 
-        if (entitylivingbaseIn == null)
-        {
+        if (entityliving == null) {
             this.targetChangeTime = 0;
             this.dataManager.set(SCREAMING, Boolean.valueOf(false));
             iattributeinstance.removeModifier(ATTACKING_SPEED_BOOST);
-        }
-        else
-        {
+        } else {
             this.targetChangeTime = this.ticksExisted;
             this.dataManager.set(SCREAMING, Boolean.valueOf(true));
 
-            if (!iattributeinstance.hasModifier(ATTACKING_SPEED_BOOST))
-            {
+            if (!iattributeinstance.hasModifier(ATTACKING_SPEED_BOOST)) {
                 iattributeinstance.applyModifier(ATTACKING_SPEED_BOOST);
             }
         }
+        return true;
     }
 
     protected void entityInit()
@@ -187,8 +197,17 @@
         this.setHeldBlockState(iblockstate);
     }
 
-    private boolean shouldAttackPlayer(EntityPlayer player)
+    // Paper start
+    private boolean shouldAttackPlayer(EntityPlayer player) {
+        boolean shouldAttack = shouldAttackPlayer_default(player);
+        EndermanAttackPlayerEvent event = new EndermanAttackPlayerEvent((Enderman) getBukkitEntity(), (Player) player.getBukkitEntity());
+        event.setCancelled(!shouldAttack);
+        return event.callEvent();
+    }
+
+    private boolean shouldAttackPlayer_default(EntityPlayer player)
     {
+        // Paper ends
         ItemStack itemstack = player.inventory.armorInventory.get(3);
 
         if (itemstack.getItem() == Item.getItemFromBlock(Blocks.PUMPKIN))
@@ -236,7 +255,7 @@
         {
             float f = this.getBrightness();
 
-            if (f > 0.5F && this.world.canSeeSky(new BlockPos(this)) && this.rand.nextFloat() * 30.0F < (f - 0.4F) * 2.0F)
+            if (f > 0.5F && this.world.canSeeSky(new BlockPos(this)) && this.rand.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && tryEscape(EndermanEscapeEvent.Reason.RUNAWAY)) // Paper
             {
                 this.setAttackTarget((EntityLivingBase)null);
                 this.teleportRandomly();
@@ -246,7 +265,7 @@
         super.updateAITasks();
     }
 
-    protected boolean teleportRandomly()
+    public boolean teleportRandomly() // protected->public Paper
     {
         double d0 = this.posX + (this.rand.nextDouble() - 0.5D) * 64.0D;
         double d1 = this.posY + (double)(this.rand.nextInt(64) - 32);
@@ -331,7 +350,7 @@
         {
             return false;
         }
-        else if (source instanceof EntityDamageSourceIndirect)
+        else if (source instanceof EntityDamageSourceIndirect && tryEscape(EndermanEscapeEvent.Reason.INDIRECT)) // Paper
         {
             for (int i = 0; i < 64; ++i)
             {
@@ -347,7 +366,7 @@
         {
             boolean flag = super.attackEntityFrom(source, amount);
 
-            if (source.isUnblockable() && this.rand.nextInt(10) != 0)
+            if (source.isUnblockable() && this.rand.nextInt(10) != 0 && tryEscape(source == DamageSource.DROWN ? EndermanEscapeEvent.Reason.DROWN : EndermanEscapeEvent.Reason.CRITICAL_HIT)) // Paper
             {
                 this.teleportRandomly();
             }
@@ -399,6 +418,12 @@
             private int aggroTime;
             private int teleportTime;
 
+            // Paper start
+            public EntityEnderman getEnderman() {
+                return enderman;
+            }
+            // Paper end
+
             public AIFindPlayer(EntityEnderman p_i45842_1_)
             {
                 super(p_i45842_1_, EntityPlayer.class, false);
@@ -467,7 +492,7 @@
                     {
                         if (this.enderman.shouldAttackPlayer((EntityPlayer)this.targetEntity))
                         {
-                            if (((EntityPlayer)this.targetEntity).getDistanceSq(this.enderman) < 16.0D)
+                            if (((EntityPlayer)this.targetEntity).getDistanceSq(this.enderman) < 16.0D && this.getEnderman().tryEscape(EndermanEscapeEvent.Reason.STARE)) // Paper
                             {
                                 this.enderman.teleportRandomly();
                             }
@@ -524,8 +549,12 @@
 
                 if (iblockstate2 != null && this.canPlaceBlock(world, blockpos, iblockstate2.getBlock(), iblockstate, iblockstate1) && net.minecraftforge.event.ForgeEventFactory.onBlockPlace(enderman, new net.minecraftforge.common.util.BlockSnapshot(world, blockpos, iblockstate2), net.minecraft.util.EnumFacing.UP).isCanceled())
                 {
-                    world.setBlockState(blockpos, iblockstate2, 3);
-                    this.enderman.setHeldBlockState((IBlockState)null);
+                    // CraftBukkit start - Place event
+                    if (!CraftEventFactory.callEntityChangeBlockEvent(this.enderman, blockpos, this.enderman.getHeldBlockState().getBlock(), this.enderman.getHeldBlockState().getBlock().getMetaFromState(this.enderman.getHeldBlockState())).isCancelled()) {
+                        world.setBlockState(blockpos, iblockstate2, 3);
+                        this.enderman.setHeldBlockState((IBlockState) null);
+                    }
+                    // CraftBukkit end
                 }
             }
 
@@ -590,8 +619,12 @@
 
                 if (EntityEnderman.CARRIABLE_BLOCKS.contains(block) && flag)
                 {
-                    this.enderman.setHeldBlockState(iblockstate);
-                    world.setBlockToAir(blockpos);
+                    // CraftBukkit start - Pickup event
+                    if (!CraftEventFactory.callEntityChangeBlockEvent(this.enderman, this.enderman.world.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()), org.bukkit.Material.AIR).isCancelled()) {
+                        this.enderman.setHeldBlockState(iblockstate);
+                        world.setBlockToAir(blockpos);
+                    }
+                    // CraftBukkit end
                 }
             }
         }
