--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySilverfish.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySilverfish.java
@@ -26,6 +26,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class EntitySilverfish extends EntityMob
 {
@@ -146,7 +147,7 @@
         if (super.getCanSpawnHere())
         {
             EntityPlayer entityplayer = this.world.getNearestPlayerNotCreative(this, 5.0D);
-            return entityplayer == null;
+            return !(entityplayer != null && !entityplayer.affectsSpawning) && entityplayer == null; // Paper - Affects Spawning API
         }
         else
         {
@@ -221,6 +222,9 @@
 
                     if (BlockSilverfish.canContainSilverfish(iblockstate))
                     {
+                        if (CraftEventFactory.callEntityChangeBlockEvent(this.entity, blockpos, Blocks.MONSTER_EGG, Block.getIdFromBlock(BlockSilverfish.getBlockById(iblockstate.getBlock().getMetaFromState(iblockstate)))).isCancelled()) {
+                            return;
+                        }
                         world.setBlockState(blockpos, Blocks.MONSTER_EGG.getDefaultState().withProperty(BlockSilverfish.VARIANT, BlockSilverfish.EnumType.forModelBlock(iblockstate)), 3);
                         this.entity.spawnExplosionParticle();
                         this.entity.setDead();
@@ -273,6 +277,9 @@
 
                                 if (iblockstate.getBlock() == Blocks.MONSTER_EGG)
                                 {
+                                    if (CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockpos1, Blocks.AIR, 0).isCancelled()) {
+                                        continue;
+                                    }
                                     if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(world, this.silverfish))
                                     {
                                         world.destroyBlock(blockpos1, true);
