--- ../src-base/minecraft/net/minecraft/block/BlockCactus.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCactus.java
@@ -9,7 +9,6 @@
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.entity.Entity;
-import net.minecraft.init.Blocks;
 import net.minecraft.util.BlockRenderLayer;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.EnumFacing;
@@ -19,6 +18,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class BlockCactus extends Block implements net.minecraftforge.common.IPlantable
 {
@@ -48,27 +48,27 @@
                 ;
             }
 
-            if (i < 3)
-            {
-                int j = ((Integer)state.getValue(AGE)).intValue();
-
-                if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, blockpos, state, true))
+            if (i < worldIn.paperConfig.cactusMaxHeight) { // Paper - Configurable growth height
                 {
-                if (j == 15)
-                {
-                    worldIn.setBlockState(blockpos, this.getDefaultState());
-                    IBlockState iblockstate = state.withProperty(AGE, Integer.valueOf(0));
-                    worldIn.setBlockState(pos, iblockstate, 4);
-                    iblockstate.neighborChanged(worldIn, blockpos, this, pos);
+                    int j = ((Integer) state.getValue(AGE)).intValue();
+
+                    if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, blockpos, state, true)) {
+                        if (j >= (byte) Block.range(3.0f, 100.0f / worldIn.spigotConfig.cactusModifier * 15.0f + 0.5f, 15.0f)) // Spigot
+                        {
+                            // worldIn.setBlockState(blockpos, this.getDefaultState());
+                            IBlockState iblockstate = state.withProperty(AGE, Integer.valueOf(0));
+                            CraftEventFactory.handleBlockGrowEvent(worldIn, blockpos.getX(), blockpos.getY(), blockpos.getZ(), this, 0);
+                            worldIn.setBlockState(pos, iblockstate, 4);
+                            iblockstate.neighborChanged(worldIn, blockpos, this, pos);
+                        } else {
+                            worldIn.setBlockState(pos, state.withProperty(AGE, Integer.valueOf(j + 1)), 4);
+                        }
+                        net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
+                    }
                 }
-                else
-                {
-                    worldIn.setBlockState(pos, state.withProperty(AGE, Integer.valueOf(j + 1)), 4);
-                }
-                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
-                }
             }
         }
+
     }
 
     public AxisAlignedBB getCollisionBoundingBox(IBlockState blockState, IBlockAccess worldIn, BlockPos pos)
@@ -123,7 +123,9 @@
 
     public void onEntityCollidedWithBlock(World worldIn, BlockPos pos, IBlockState state, Entity entityIn)
     {
+        CraftEventFactory.blockDamage = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
         entityIn.attackEntityFrom(DamageSource.CACTUS, 1.0F);
+        CraftEventFactory.blockDamage = null;
     }
 
     public IBlockState getStateFromMeta(int meta)
