--- ../src-base/minecraft/net/minecraft/block/BlockIce.java
+++ ../src-work/minecraft/net/minecraft/block/BlockIce.java
@@ -19,6 +19,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class BlockIce extends BlockBreakable
 {
@@ -86,6 +87,9 @@
 
     protected void turnIntoWater(World worldIn, BlockPos pos)
     {
+        if (CraftEventFactory.callBlockFadeEvent(worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), worldIn.provider.doesWaterVaporize() ? Blocks.AIR : Blocks.WATER).isCancelled()) {
+            return;
+        }
         if (worldIn.provider.doesWaterVaporize())
         {
             worldIn.setBlockToAir(pos);
