--- ../src-base/minecraft/net/minecraft/item/ItemFireball.java
+++ ../src-work/minecraft/net/minecraft/item/ItemFireball.java
@@ -11,6 +11,7 @@
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class ItemFireball extends Item
 {
@@ -38,6 +39,12 @@
             {
                 if (worldIn.getBlockState(pos).getMaterial() == Material.AIR)
                 {
+                    if (CraftEventFactory.callBlockIgniteEvent(worldIn, pos.getX(), pos.getY(), pos.getZ(), org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, player).isCancelled()) {
+                        if (!player.capabilities.isCreativeMode) {
+                            itemstack.shrink(1);
+                        }
+                        return EnumActionResult.PASS;
+                    }
                     worldIn.playSound((EntityPlayer)null, pos, SoundEvents.ITEM_FIRECHARGE_USE, SoundCategory.BLOCKS, 1.0F, (itemRand.nextFloat() - itemRand.nextFloat()) * 0.2F + 1.0F);
                     worldIn.setBlockState(pos, Blocks.FIRE.getDefaultState());
                 }
