--- ../src-base/minecraft/net/minecraft/network/play/server/SPacketTitle.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/SPacketTitle.java
@@ -7,6 +7,9 @@
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayClient;
 import net.minecraft.util.text.ITextComponent;
+
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.chat.ComponentSerializer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
@@ -18,6 +21,19 @@
     private int displayTime;
     private int fadeOutTime;
 
+    // Paper start
+    public BaseComponent[] components;
+
+    public SPacketTitle(SPacketTitle.Type typeIn, BaseComponent[] components, int fadeInTimeIn, int displayTimeIn, int fadeOutTimeIn)
+    {
+        this.type = typeIn;
+        this.components = components;
+        this.fadeInTime = fadeInTimeIn;
+        this.displayTime = displayTimeIn;
+        this.fadeOutTime = fadeOutTimeIn;
+    }
+    // Paper end
+
     public SPacketTitle()
     {
     }
@@ -64,7 +80,13 @@
 
         if (this.type == SPacketTitle.Type.TITLE || this.type == SPacketTitle.Type.SUBTITLE || this.type == SPacketTitle.Type.ACTIONBAR)
         {
-            buf.writeTextComponent(this.message);
+            // Paper start
+            if(this.components != null){
+                buf.writeString(ComponentSerializer.toString(components));
+            }else{
+                buf.writeTextComponent(this.message);
+            }
+            // Paper end
         }
 
         if (this.type == SPacketTitle.Type.TIMES)
